mpc-core/src/protocols/rep3/yao.rs

```diff
@@ -0,0 +1,481 @@
+            network.send(garbler_id, val.as_block().as_ref())?;
+        }
+        for val in input.evaluator_wires.iter() {
+            network.send(PartyID::ID0, val.as_block().as_ref())?;
+        }
+
+        Ok(())
+    }
+
+    /// Samples a random delta
+    pub fn random_delta<R: Rng + CryptoRng>(rng: &mut R) -> WireMod2 {
+        WireMod2::rand_delta(rng, 2)
+    }
+
+    #[cfg(test)]
+    fn u16_bits_to_field<F: PrimeField>(bits: Vec<u16>) -> IoResult<F> {
+        let mut res = BigUint::zero();
+        for bit in bits.iter().rev() {
+            assert!(*bit < 2);
+            res <<= 1;
+            res += *bit as u64;
+        }
+
+        if res >= F::MODULUS.into() {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                "Invalid field element",
+            ));
+        }
+        Ok(F::from(res))
+    }
+
+    /// Converts bits into a field element
+    pub fn bits_to_field<F: PrimeField>(bits: Vec<bool>) -> IoResult<F> {
+        let mut res = BigUint::zero();
+        for bit in bits.iter().rev() {
+            res <<= 1;
+            res += *bit as u64;
+        }
+        if res >= F::MODULUS.into() {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                "Invalid field element",
+            ));
+        }
+        Ok(F::from(res))
+    }
+
+    fn biguint_to_bits(input: BigUint, n_bits: usize) -> Vec<bool> {
+        let mut res = Vec::with_capacity(n_bits);
+        let mut bits = 0;
+        for mut el in input.to_u64_digits() {
+            for _ in 0..64 {
+                res.push(el & 1 == 1);
+                el >>= 1;
+                bits += 1;
+                if bits == n_bits {
+                    break;
+                }
+            }
+        }
+        res.resize(n_bits, false);
+        res
+    }
+
+    fn field_to_bits_as_u16<F: PrimeField>(field: F) -> Vec<u16> {
+        let n_bits = F::MODULUS_BIT_SIZE as usize;
+        let bigint: BigUint = field.into();
+
+        Self::biguint_to_bits_as_u16(&bigint, n_bits)
+    }
+
+    fn biguint_to_bits_as_u16(input: &BigUint, n_bits: usize) -> Vec<u16> {
+        let mut res = Vec::with_capacity(n_bits);
+        let mut bits = 0;
+        for mut el in input.to_u64_digits() {
+            for _ in 0..64 {
+                res.push((el & 1) as u16);
+                el >>= 1;
+                bits += 1;
+                if bits == n_bits {
+                    break;
+                }
+            }
+        }
+        res.resize(n_bits, 0);
+        res
+    }
+
+    /// Encode a wire, producing the zero wire as well as the encoded value.
+    pub fn encode_wire<R: Rng + CryptoRng>(
+        rng: &mut R,
+        delta: &WireMod2,
+        val: u16,
+    ) -> (WireMod2, WireMod2) {
+        let zero = WireMod2::rand(rng, 2);
+        let enc = zero.plus(&delta.cmul(val));
+        (zero, enc)
+    }
```

Comment:
This specialized for the arithmetic garbling already right? `val` can never be non {0,1} no? also if val == 0, does this not return (zero, zero)?

Reply:
Ok this does not produce (zero_wire, one_wire) but (garbler_wire, evaluator_wire).

Reply:
Indeed, val is only 0 or 1, but it sometimes need to be u16 to be compatible with the traits from fancygarble.
garbler_wire is the zero_wire, and evaluator wire is the zero_wire xor delta * x (where x is 0 or 1)

Reply:
ok then lets leave it like it is atm